edgelet/aziot-edged/src/product_info.rs

```diff
@@ -0,0 +1,194 @@
+pub(crate) struct ProductInfo {
+    product: Vec<Product>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct Product {
+    name: String,
+    version: String,
+    comment: Option<String>,
+}
+
+#[inline]
+fn read_utf8(v: Vec<u8>) -> io::Result<String> {
+    Ok(String::from_utf8(v)
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?
+        .trim()
+        .to_owned())
+}
+
+#[inline]
+fn read_output(out: Output) -> io::Result<String> {
+    match out.status.code() {
+        Some(0) => read_utf8(out.stdout),
+        Some(code) => Err(io::Error::from_raw_os_error(code)),
+        _ => Err(io::Error::from(io::ErrorKind::Other)),
+    }
+}
+
+#[inline]
+fn strip_quotes(s: &str) -> io::Result<&str> {
+    s.strip_prefix('\"')
+        .and_then(|s| s.strip_suffix('\"'))
+        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "mismatched quotes"))
+}
+
+impl ProductInfo {
+    pub fn try_load(p: impl AsRef<Path>) -> io::Result<Self> {
+        let bytes = fs::read(p)?;
+
+        toml::de::from_slice(&bytes).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    }
+
+    pub fn from_system() -> io::Result<Self> {
+        let mut product = vec![];
+
+        // NOTE: operating system
+        product.push({
+            let b = io::BufReader::new(
+                fs::File::open("/etc/os-release")
+                    .or_else(|_| fs::File::open("/usr/lib/os-release"))?,
+            );
+
+            let mut name = None;
+            let mut version = None;
+            let mut comment = None;
+
+            for line in b.lines() {
+                match line?.trim().split_once("=") {
+                    Some(("ID", value)) => name = Some(strip_quotes(value)?.to_owned()),
+                    Some(("VERSION_ID", value)) => version = Some(strip_quotes(value)?.to_owned()),
+                    Some(("PRETTY_NAME", value)) => comment = Some(strip_quotes(value)?.to_owned()),
+                    _ => (),
+                }
+            }
+
+            Product {
+                name: name
+                    .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "os-release: ID"))?,
+                version: version.ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::NotFound, "os-release: VERSION_ID")
+                })?,
+                comment,
+            }
+        });
+
+        // NOTE: kernel
+        product.push(Product {
+            name: Command::new("uname")
+                .arg("-s")
+                .output()
+                .and_then(read_output)?,
+            version: Command::new("uname")
+                .arg("-r")
+                .output()
+                .and_then(read_output)?,
+            comment: Some(
+                Command::new("uname")
+                    .arg("-v")
+                    .output()
+                    .and_then(read_output)?,
+            ),
+        });
+
+        // NOTE: device
+        product.push(Product {
```

Comment:
Does this exist on all hardware? I'm finding pages on the internet indicating it might not work for ARM hardware, though I haven't tested it myself.

In general, I think it would be good to allow reading these files to fail instead of bubbling the error up and failing the entire fn, since some info is better than no info. That's what `aziotctl check`'s `OsInfo` does when parsing `/etc/os-release`, for example.